### PR TITLE
[5.2] Added link class selection to TinyMCE dialog

### PR DESCRIPTION
**Pull request by RickR2H**
Pull Request for Issue # .

### Summary of Changes
Add a class selector to the "create link" dialog. The selected class will be added to the link. This can be very useful in styling links on the website. Adding Bootstrap button classes will convert the link to a button.

### Testing Instructions
- Go to the tinyMCE configuration and op the "Advanced tab"

tinyMCE-config

- Add some classes to use in the dropdown
- Open an article and create a link. At the bottom the configured classes can be selected.

tinyMCE-dialog

### Actual result BEFORE applying this Pull Request
Adding classed only possible in the source code editor.

### Expected result AFTER applying this Pull Request
Classes can be selected in the link dialog.

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: (will be created after merging the PR)
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: plugins/editors/tinymce/tinymce.xml [hunk 1]
```diff
@@ -57,6 +57,36 @@
...
+					label="PLG_TINY_FIELD_LINK_CLASSES_LIST_LABEL"
+					description=""
+					required="false"
+					multiple="true">
+					<form>
+						<field
+							name="class_name"
+							label="PLG_TINY_FIELD_LINK_CLASS_NAME_LABEL"
+							description="PLG_TINY_FIELD_LINK_CLASS_NAME_DESC"
+							type="text"
```
Comment (brianteeman) [inline plugins/editors/tinymce/tinymce.xml]:
```suggestion
							name="class_name"
							type="text"
							label="PLG_TINY_FIELD_LINK_CLASS_NAME_LABEL"
							description="PLG_TINY_FIELD_LINK_CLASS_NAME_DESC"
```

File: plugins/editors/tinymce/tinymce.xml [hunk 2]
```diff
@@ -57,6 +57,36 @@
...
+							required="true"
+							filter="string"
+							default=""
+						/>
+						<field
+							name="class_list"
+							label="PLG_TINY_FIELD_LINK_CLASS_LIST_LABEL"
+							description="PLG_TINY_FIELD_LINK_CLASS_LIST_DESC"
+							type="text"
```
Comment (brianteeman) [inline plugins/editors/tinymce/tinymce.xml]:
```suggestion
							name="class_list"
							type="text"
							label="PLG_TINY_FIELD_LINK_CLASS_LIST_LABEL"
							description="PLG_TINY_FIELD_LINK_CLASS_LIST_DESC"
```

File: administrator/language/en-GB/plg_editors_tinymce.ini [hunk 3]
```diff
@@ -44,6 +44,12 @@ PLG_TINY_FIELD_HTMLWIDTH_LABEL="HTML Width"
 PLG_TINY_FIELD_LABEL_ADVANCEDPARAMS="Advanced"
 PLG_TINY_FIELD_LANGCODE_LABEL="Language Code"
 PLG_TINY_FIELD_LANGSELECT_LABEL="Automatic Language Selection"
+PLG_TINY_FIELD_LINK_CLASS_LIST_DESC="The classes that wil be added to the link. Separate individual classes by a space (example: btn btn-primary). Only valid class names are allowed."
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_editors_tinymce.ini]: spelling wil ==> will
But I think it is better to say "may"
```suggestion
PLG_TINY_FIELD_LINK_CLASS_LIST_DESC="The classes that may be added to the link. Separate individual classes by a space (example: btn btn-primary). Only valid class names are allowed."
```

File: administrator/language/en-GB/plg_editors_tinymce.ini [hunk 4]
```diff
@@ -44,6 +44,12 @@ PLG_TINY_FIELD_HTMLWIDTH_LABEL="HTML Width"
...
+PLG_TINY_FIELD_LINK_CLASS_LIST_LABEL="Class List"
+PLG_TINY_FIELD_LINK_CLASS_NAME_DESC="The name shown in the link class dropdown in TinyMCE."
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_editors_tinymce.ini]: is this description needed. Is it not obvious?

**Comment by brianteeman** [comment]
works great. I never would have found it on the advanced tab

**Comment by RickR2H** [comment]
> works great. I never would have found it on the advanced tab

I could move it to the first "Plugin" tab? Would that be a better option?

**Comment by brianteeman** [comment]
> > works great. I never would have found it on the advanced tab
> 
> I could move it to the first "Plugin" tab? Would that be a better option?

I would prefer that. It is where I added all the other tinymce customisations in the past

**Comment by RickR2H** [comment]
@brianteeman I added it to the first tab at the bottom. I also removed the description of the single fields and added an overall description.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 0f4bbcb5c3139d39519ab516f502d02a5b287845<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43260">issues.joomla.org/tracker/joomla-cms/43260</a>.</sub>

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php [hunk 5]
```diff
@@ -382,6 +382,19 @@ public function display(string $name, string $content = '', array $attributes =
         // Merge the two toolbars for backwards compatibility
         $toolbar = array_merge($toolbar1, $toolbar2);
 
+        // Load the link classes list
+        $linksClassesList = $extraOptions->link_classes_list;
+        $linkClasses      = [];
+
+        if ($linksClassesList) {
+            $linkClasses      = [['title' => TEXT::_('PLG_TINY_FIELD_LINK_CLASS_NONE'), 'value' => '']];
```
Comment (Quy) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]:
```suggestion
            $linkClasses = [['title' => TEXT::_('PLG_TINY_FIELD_LINK_CLASS_NONE'), 'value' => '']];
```

File: plugins/editors/tinymce/tinymce.xml [hunk 6]
```diff
@@ -57,6 +57,7 @@
 					label="PLG_TINY_FIELD_HTMLWIDTH_LABEL"
 					default=""
 				/>
+
 			</fieldset>
```
Comment (Quy) [inline plugins/editors/tinymce/tinymce.xml]:
```suggestion
				/>
			</fieldset>
```

File: plugins/editors/tinymce/forms/setoptions.xml [hunk 7]
```diff
@@ -403,5 +403,34 @@
...
+					label="PLG_TINY_FIELD_LINK_CLASS_LIST_LABEL"
+					required="true"
+					filter="string"
+					validate="CssIdentifier"
+					default=""
+				/>
+			</form>
+		</field>
+
 	</fieldset>
```
Comment (Quy) [inline plugins/editors/tinymce/forms/setoptions.xml]:
```suggestion
		</field>
	</fieldset>
```

**Comment by cybersalt** [comment]
I have tested this item :white_check_mark: successfully on 96ee302326f02f7d499cd176f502f01c25195aee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43260">issues.joomla.org/tracker/joomla-cms/43260</a>.</sub>

**Comment by RickR2H** [comment]
@brianteeman could you please do one more test? Code is fine now...

**Comment by Quy** [comment]
@RickR2H Please rebase for v5.2 as this will be considered as a new feature. Only bug fixes will be merged in v5.1. Thank you!

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php [hunk 8]
```diff
@@ -382,6 +382,23 @@ public function display(string $name, string $content = '', array $attributes =
         // Merge the two toolbars for backwards compatibility
         $toolbar = array_merge($toolbar1, $toolbar2);
 
+        // Set default classes to empty
+        $linkClasses      = [];
```
Comment (Quy) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]:
```suggestion
        $linkClasses = [];
```

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43260">issues.joomla.org/tracker/joomla-cms/43260</a>.</sub>

**Comment by Quy** [comment]
Restored tests as latest changes are cs only.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43260">issues.joomla.org/tracker/joomla-cms/43260</a>.</sub>

**Comment by pe7er** [comment]
Thanks @RickR2H !

**Pull request merged**